superset/assets/src/SqlLab/actions/sqlLab.js

```diff
@@ -132,7 +139,23 @@ export function startQuery(query) {
 }
 
 export function querySuccess(query, results) {
-  return { type: QUERY_SUCCESS, query, results };
+  return function (dispatch) {
+    // table preview queries have `sqlEditorId` set to the string 'null'; those
+    // shouldn't be synced to the tab state view since they're not the main query
+    const isDataPreview = results.query && results.query.sqlEditorId === 'null';
+    const sync = (!isDataPreview && isFeatureEnabled(FeatureFlag.SQLLAB_BACKEND_PERSISTENCE))
+      ? SupersetClient.put({
+          endpoint: encodeURI(`/tabstateview/${results.query.sqlEditorId}`),
+          postPayload: { query_id: results.query_id },
+        })
+      : Promise.resolve();
+
+    return sync
+      .then(() => dispatch({ type: QUERY_SUCCESS, query, results }))
+      .catch(() =>
+        dispatch(addDangerToast(t('An error occurred while updating tab query'))),
```

Comment:
"Saving the query" conflicts with explicitly saving a query (where it asks for a name, description, and it ends in "Saved Queries"). I agree we can make it better — I like error messages that are clearly actionable, so maybe something like:

> An error occurred while storing the tab state in the backend. To avoid losing your changes, please save your query using the "Save Query" button.